x/blob/types/blob_tx.go

```diff
@@ -0,0 +1,118 @@
+package types
+
+import (
+	"bytes"
+
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	"github.com/celestiaorg/nmt/namespace"
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+)
+
+// ProcessedBlobTx caches the unmarshalled result of the BlobTx
+type ProcessedBlobTx struct {
+	Blobs []tmproto.Blob
+	Tx    []byte
+}
+
+// NewBlob creates a new coretypes.Blob from the provided data after performing
+// basic stateless checks over it.
+func NewBlob(ns namespace.ID, blob []byte) (*tmproto.Blob, error) {
+	err := ValidateBlobNamespaceID(ns)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(blob) == 0 {
+		return nil, ErrZeroBlobSize
+	}
+
+	return &tmproto.Blob{
+		NamespaceId:  ns,
+		Data:         blob,
+		ShareVersion: uint32(appconsts.ShareVersionZero),
+	}, nil
+}
+
+// ProcessBlobTx performs stateless checks on the BlobTx to ensure that the
+// blobs attached to the transaction are valid. During this process, it
+// separates the blobs from the MsgPayForBlob, which are returned in the
+// ProcessedBlobTx.
+//
+// NOTE: ProcessBlobTx does not call the ValidateBasic method on either the
+// transaction or messages in that transaction.
+func ProcessBlobTx(txcfg client.TxEncodingConfig, bTx tmproto.BlobTx) (ProcessedBlobTx, error) {
```

Comment:
So with this, it's already possible to submit multiple other random sdk.Msg's in the `BlobTx` and have multiple PFBs/Blobs?

Reply:
that's a good question, and I did some testing and it wasn't working. 

That wasn't an explicit goal of this PR, so instead of leaving ambiguous I added an extra temporary explicit check to prevent this. https://github.com/celestiaorg/celestia-app/pull/1089/commits/5978eef3c6b3cb7a4beb9d89a6081735e5f84c17

Note that this additional check does not change actually enforce that as a block validity rule! Its only a default. If a block producer does this now, the transaction will fail, but pfb and its blob will still get included. This is not unlike submitting a pfb from an account that doesn't have enough funds to pay for gas, the blob will still be included. https://github.com/celestiaorg/celestia-app/issues/890#issuecomment-1344984604